config/env_sample.json

```diff
@@ -239,25 +239,25 @@
         "canvas": {
             "query": [
                 "SELECT 'canvas' AS resource_type, ",
-                "CAST(SUBSTR(JSON_EXTRACT_SCALAR(event, '$.object.id'), 35) AS STRING) AS resource_id, ",
-                "CAST(SUBSTR(JSON_EXTRACT_SCALAR(event, '$.membership.member.id'), 29) AS INT64) AS user_id, ",
-                "cast(null as string) AS user_login_name,",
-                "CAST(SUBSTR(JSON_EXTRACT_SCALAR(event, '$.group.id'), 31) AS INT64) AS course_id, ",
+                "CAST(SUBSTR(object.id, -17) AS STRING) AS resource_id, ",
```

Comment:
Looking at this, I think we might want to replace these with something more like

`REGEXP_EXTRACT(object.id, r".*[/:](.*)")`

This would return anything after the last : or / character which seems like it would more reliably capture just the id piece we want. -17 would work for Canvas, but would fail for other types. Using REGEXP_EXTRACT should work for everything.

Reply:
Yeah, I don't know what would be more efficient but I didn't find any string method for BigQuery to do that, and did see a lot of regex examples.

Reply:
TL;DR: Using `REGEXP_EXTRACT` seems to be the way to go, unfortunately.

Yes.  I know some languages have functions that can search for one or more characters, from the beginning or end of the string.  Like (this is pseudo-code), `findChars('12:34/45', ':/', 1)` would return the index of `:` if searching from first character (`1`) or `findChars('12:34/45', ':/', -1)` would return the index of `/` if searching from last character (`-1`).  I thought Python or JavaScript have that feature.

Std. SQL's `INSTR` can't search for multiple characters that way.  `REGEXP_INSTR` could.  It would mean using a regex, but just a simple one like `[:/]`.  I see that the documentation for `REGEXP_INSTR` says it can search from position `-1`.  **SPOILER ALERT:**  The documentation is wrong.  That gives an error.

I looked into making a user-defined function (UDF) or procedure for this.  That's too convoluted.

I'm disappointed with BigQuery's standard SQL right now.

Reply:
Thanks for the suggestions. I have made the changes accordingly.